File: bin/zxdb/client/file_util.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "garnet/bin/zxdb/client/file_util.h"

namespace zxdb {

fxl::StringView ExtractLastFileComponent(fxl::StringView path) {
  size_t last_slash = path.rfind('/');
  if (last_slash == std::string::npos)
    return path;
  return path.substr(last_slash + 1);
}

bool IsPathAbsolute(const std::string& path) {
  return !path.empty() && path[0] == '/';
}

std::string CatPathComponents(const std::string& first,
                              const std::string& second) {
  // Second component shouldn't begin with a slash.
  FXL_DCHECK(second.empty() || second[0] != '/');

  std::string result;
  result.reserve(first.size() + second.size() + 1);
  result.append(first);

  if (!first.empty() && !second.empty() && first.back() != '/')
    result.push_back('/');
  result.append(second);

  return result;
}

}  // namespace zxdb

File: tools/botanist/fileutil.go
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package botanist

import (
	"bytes"
	"context"
	"fmt"
	"path/filepath"
	"time"

	"go.fuchsia.dev/fuchsia/tools/botanist/constants"
	"go.fuchsia.dev/fuchsia/tools/lib/osmisc"
	"go.fuchsia.dev/fuchsia/tools/lib/retry"
	"go.fuchsia.dev/fuchsia/tools/net/tftp"
)

// FetchAndCopyFile fetches a remote file via TFTP from a given node, and
// writes it to an output directory.
func FetchAndCopyFile(ctx context.Context, t tftp.Client, path, name, outDir string) error {
	return retry.Retry(ctx, retry.WithMaxAttempts(retry.NewConstantBackoff(time.Second*5), 3), func() error {
		var err error
		var reader *bytes.Reader
		for {
			reader, err = t.Read(ctx, path)
			switch err {
			case nil:
			case tftp.ErrShouldWait:
				time.Sleep(time.Second)
				continue
			default:
				return fmt.Errorf("%s %s: %s", constants.FailedToReceiveFileMsg, path, err)
			}
			break
		}
		outputFile := filepath.Join(outDir, name)
		w, err := osmisc.CreateFile(outputFile)
		if err != nil {
			return fmt.Errorf("failed to create file: %v", err)
		}
		defer w.Close()
		_, err = reader.WriteTo(w)
		return err
	}, nil)
}
